src/portfft/dispatcher/global_dispatcher.hpp

```diff
@@ -351,16 +356,32 @@ struct committed_descriptor<Scalar, Domain>::run_kernel_struct<Dir, LayoutIn, La
       for (Idx num_transpose = num_transposes - 1; num_transpose > 0; num_transpose--) {
         event = detail::transpose_level<Scalar, Domain>(
             kernels.at(static_cast<std::size_t>(num_transpose) + static_cast<std::size_t>(num_factors)),
-            static_cast<const Scalar*>(desc.scratch_ptr_1.get()), desc.scratch_ptr_2.get(), factors_and_scan,
-            committed_size, static_cast<Idx>(max_batches_in_l2), n_transforms, static_cast<IdxGlobal>(i), num_transpose,
-            num_factors, 0, desc.queue, desc.scratch_ptr_1, desc.scratch_ptr_2, {event});
+            desc.scratch_ptr_1.get(), desc.scratch_ptr_2.get(), factors_and_scan, committed_size,
+            static_cast<Idx>(max_batches_in_l2), n_transforms, static_cast<IdxGlobal>(i), num_factors, 0, desc.queue,
+            {event}, storage);
         event.wait();
+        if (storage == complex_storage::SPLIT_COMPLEX) {
+          event = detail::transpose_level<Scalar, Domain>(
+              kernels.at(static_cast<std::size_t>(num_transpose) + static_cast<std::size_t>(num_factors)),
+              desc.scratch_ptr_1.get() + imag_offset, desc.scratch_ptr_2.get() + imag_offset, factors_and_scan,
+              committed_size, static_cast<Idx>(max_batches_in_l2), n_transforms, static_cast<IdxGlobal>(i), num_factors,
+              0, desc.queue, {event}, storage);
+          event.wait();
+        }
+        desc.scratch_ptr_1.swap(desc.scratch_ptr_2);
       }
       event = detail::transpose_level<Scalar, Domain>(
-          kernels.at(static_cast<std::size_t>(num_factors)), static_cast<const Scalar*>(desc.scratch_ptr_1.get()), out,
-          factors_and_scan, committed_size, static_cast<Idx>(max_batches_in_l2), n_transforms,
-          static_cast<IdxGlobal>(i), 0, num_factors, 2 * static_cast<IdxGlobal>(i) * committed_size + output_offset,
-          desc.queue, desc.scratch_ptr_1, desc.scratch_ptr_2, {event});
+          kernels.at(static_cast<std::size_t>(num_factors)), desc.scratch_ptr_1.get(), out, factors_and_scan,
+          committed_size, static_cast<Idx>(max_batches_in_l2), n_transforms, static_cast<IdxGlobal>(i), num_factors,
+          vec_size * static_cast<IdxGlobal>(i) * committed_size + output_offset, desc.queue, {event}, storage);
+      if (storage == complex_storage::SPLIT_COMPLEX) {
+        event.wait();
```

Comment:
You needn't wait on the real transpositions to start the `imag` transpositions.
Same for the previous ones as well.

Reply:
Checking again waits can be removed. Done.